### PR TITLE
X cell sum

### PR DESCRIPTION
**Pull request by drkwjhnsn**
Displays column sums in html table footer.

### CODE REVIEW

File: client/js/table-model.js [hunk 1]
```diff
@@ -16,6 +17,27 @@ class TableModel {
   _setValue(location, value) {
     this.data[this._getCellId(location)] = value;
   }
+
+  findSum(column) {
+    let sum = 0;
+    for (let location in this.data) {
+      let value = parseInt(this.data[location], 10);
+      if (location.indexOf(`${column}:`) !== -1 &&
```
Comment (speschellk) [inline client/js/table-model.js]: This is a fine implementation, BUT can you think of a way to use .filter() to produce the same filtering effect?

File: client/js/table-view.js [hunk 2]
```diff
@@ -73,6 +75,22 @@ class TableView {
     this.sheetBodyEl.appendChild(fragment);
   }
 
+  renderTableFooter() {
+    const fragment = document.createDocumentFragment();
```
Comment (speschellk) [inline client/js/table-view.js]: Even better if you don't need to make a new document fragment to render the table footer. Check out the renderTableHeader function and how it's not using the fragment technique.

File: client/js/test/table-model.test.js [hunk 3]
```diff
@@ -15,4 +15,23 @@ describe('table-model', () => {
     // inspect the resulting state
     expect(model.getValue(location)).toBe('foo');
   });
+
+  it('can add columns, store sums and return sums', () => {
```
Comment (speschellk) [inline client/js/test/table-model.test.js]: Make a separate test for each functionality you want to test. Testing multiple functionalities in one test leads to imperfect and confusing tests; it's much harder to debug.

Additionally, your explanation ('can add columns, store sums') must describe in detail what functionality you are testing. "it can add columns" --- when? At all? When there are only numbers as inputs? When there are non-number inputs only? When there is a mix of number and non-number inputs?

Add several more tests to examine edge cases.

File: client/js/test/table-view.test.js [hunk 4]
```diff
@@ -10,6 +10,26 @@ describe('table-view', () => {
     const html = fs.readFileSync(fixturePath, 'utf8');
     document.documentElement.innerHTML = html;
 
+  describe('footer', () => {
+    it('displays column sum in table footer', () => {
```
Comment (speschellk) [inline client/js/test/table-view.test.js]: Yes, this is a more specific test explanation.

**Review by speschellk — CHANGES_REQUESTED** [review]

File: client/js/table-model.js [hunk 5]
```diff
@@ -16,6 +17,26 @@ class TableModel {
   _setValue(location, value) {
     this.data[this._getCellId(location)] = value;
   }
+
+  findSum(column) {
+    let keys = Object.keys(this.data);
+
```
Comment (speschellk) [inline client/js/table-model.js]: No need for extra blank lines here. Generally, just use an extra blank line if you want to separate one block from another, or a title line from the block underneath it, for instance.

File: client/js/test/fixtures/sheet-container.html [hunk 6]
```diff
@@ -12,5 +12,11 @@
 
       </tr>
     </tbody>
+
+    <tfoot>
+      <tr>
+
```
Comment (speschellk) [inline client/js/test/fixtures/sheet-container.html]: No need for extra blank lines between opening and closing tag. It breaks up the visual flow in a way that suggests that you WERE going to write code there and then forgot to.

File: public/index.html [hunk 7]
```diff
@@ -19,6 +19,12 @@
 
           </tr>
         </tbody>
+
+        <tfoot>
+          <tr>
+
```
Comment (speschellk) [inline public/index.html]: See above comment about lines

**Review by speschellk — CHANGES_REQUESTED** [review]
